witnesslint.py

```diff
@@ -0,0 +1,1128 @@
+    )
+    parser.add_argument(
+        "--checkProgram",
+        help="Perform some additional checks involving the program file. "
+        "Better left disabled for big witnesses. This option is "
+        "implicitly used when a program is given via the --program option.",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--checkCallstack",
+        help="Check whether transitions specified via enterFunction "
+        "and returnFromFunction are consistent. "
+        "Better left disabled for big witnesses.",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--ignoreSelfLoops",
+        help="Produce no warnings when encountering "
+        "edges that represent a self-loop.",
+        action="store_true",
+    )
+    return parser
+
+
+def create_logger(loglevel):
+    """Initializes the logger instances used in the linter."""
+    pos_logger = logging.getLogger("with_position")
+    if not pos_logger.hasHandlers():
+        pos_handler = logging.StreamHandler()
+        pos_formatter = logging.Formatter("%(levelname)-8s: line %(line)s: %(message)s")
+        pos_handler.setFormatter(pos_formatter)
+        pos_logger.addHandler(pos_handler)
+    pos_logger.setLevel(loglevel)
+
+    no_pos_logger = logging.getLogger("without_position")
+    if not no_pos_logger.hasHandlers():
+        no_pos_handler = logging.StreamHandler()
+        no_pos_formatter = logging.Formatter("%(levelname)-8s: %(message)s")
+        no_pos_handler.setFormatter(no_pos_formatter)
+        no_pos_logger.addHandler(no_pos_handler)
+    no_pos_logger.setLevel(loglevel)
+
+
+class WitnessLint:
+    """
+    Contains methods that check different parts of a witness for consistency
+    with the witness format as well as some utility methods for this purpose.
+    The lint() method checks the whole witness.
+    """
+
+    def __init__(
+        self, witness, program, check_program, check_callstack, ignore_self_loops
+    ):
+        self.witness = witness
+        self.program_info = None
+        if program is not None:
+            self.collect_program_info(program)
+        self.check_program = check_program
+        self.check_callstack = check_callstack
+        self.ignore_self_loops = ignore_self_loops
+        self.witness_type = None
+        self.sourcecodelang = None
+        self.producer = None
+        self.specifications = set()
+        self.programfile = None
+        self.programhash = None
+        self.architecture = None
+        self.creationtime = None
+        self.entry_node = None
+        self.node_ids = set()
+        self.sink_nodes = set()
+        self.transition_sources = set()
+        self.defined_keys = dict()
+        self.used_keys = set()
+        self.threads = dict()
+        self.transitions = dict()
+        self.violation_witness_only = set()
+        self.correctness_witness_only = set()
+        self.check_existence_later = set()
+        self.check_later = list()
```

Comment:
I also considered adding a separate witness class but am unsure of how useful it would be.
The linter would either just directly access the attributes of the witness instead of its own, or delegate some tasks to the witness that are currently done in the linter, no? Basically, it seems to me that this would only move around code but would not really simplify the process.

Reply:
Yes, it would move around the code and give more structure to it.
It would decouple the data, i.e., the witness, and the process, i.e. linting.

IMO it would make the code more readable.

Reply:
Alright, then I will make a `Witness` class and put it into its own module together with the constants.